doc/src/devdocs/types.md

```diff
@@ -315,7 +315,129 @@ Ptr{Void} @0x00007f5998a04370
 
 so `Tuple == Tuple{Vararg{Any}}` is indeed the primary type.
 
-## Introduction to the internal machinery: `jltypes.c`
+## Diagonal types
+
+Consider the type `Tuple{T,T} where T`.
+A method with this signature would look like:
+
+```julia
+f(x::T, y::T) where T = ...
+```
+
+According to the usual interpretation of a `UnionAll` type, this `T` ranges over all
+types, including `Any`, so this type should be equivalent to `Tuple{Any,Any}`.
+However, this interpretation causes some practical problems.
+
+First, a value of `T` needs to be available inside the method definition.
+For a call like `f(1, 1.0)`, it's not clear what `T` should be.
+It could be `Union{Int,Float64}`, or perhaps `Real`.
+Intuitively, we expect the declaration `x::T` to mean `T === typeof(x)`.
+To make sure that invariant holds, we need `typeof(x) === typeof(y) === T` in this method.
+That implies the method should only be called for arguments of the exact same type.
+
+It turns out that being able to dispatch on whether two values have the same type
+is very useful (this is used by the promotion system for example), so we have
+multiple reasons to want a different interpretation of `Tuple{T,T} where T`.
+To make this work we add the following rule to subtyping: if a variable occurs
+more than once in covariant position, it is restricted to ranging over only concrete
+types.
+("Covariant position" means that only `Tuple` and `Union` types occur between an
+occurrence of a variable and the `UnionAll` type that introduces it.)
+Such variables are called "diagonal variables" or "concrete variables".
```

Comment:
For technical writing, there is a strong countermovement, since the exact contents of the quotations is so frequently of great significance. I think that we should adopt the British rule for Julia docs.